File: layouts/max.c
/*
 * max.c - max layout
 *
 * Copyright © 2007-2008 Julien Danjou <julien@danjou.info>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

#include "tag.h"
#include "screen.h"
#include "client.h"
#include "layouts/max.h"

extern awesome_t globalconf;

static void
layout_fmax(int screen, bool fs)
{
    client_t *c;
    area_t area = screen_area_get(screen,
                                  fs ? NULL : &globalconf.screens[screen].wiboxes,
                                  fs ? NULL : &globalconf.screens[screen].padding,
                                  true);

    for(c = globalconf.clients; c; c = c->next)
        if(IS_TILED(c, screen))
        {
            area.width -= 2 * c->border;
            area.height -= 2 * c->border;
            client_resize(c, area, false);
            area.width += 2 * c->border;
            area.height += 2 * c->border;
        }

    if(IS_TILED(globalconf.screens[screen].client_focus, screen))
        client_raise(globalconf.screens[screen].client_focus);
}

void
layout_max(int screen)
{
    return layout_fmax(screen, false);
}

void
layout_fullscreen(int screen)
{
    return layout_fmax(screen, true);
}

// vim: filetype=c:expandtab:shiftwidth=4:tabstop=8:softtabstop=4:encoding=utf-8:textwidth=80

File: lib/awful/layout/suit/max.lua
---------------------------------------------------------------------------
--- Maximized and fullscreen layouts module for awful
--
-- @author Julien Danjou &lt;julien@danjou.info&gt;
-- @copyright 2008 Julien Danjou
-- @module awful.layout
---------------------------------------------------------------------------

-- Grab environment we need
local pairs = pairs

local max = {}

--- The max layout layoutbox icon.
-- @beautiful beautiful.layout_max
-- @param surface
-- @see gears.surface

--- The fullscreen layout layoutbox icon.
-- @beautiful beautiful.layout_fullscreen
-- @param surface
-- @see gears.surface

local function fmax(p, fs)
    -- Fullscreen?
    local area
    if fs then
        area = p.geometry
    else
        area = p.workarea
    end

    for _, c in pairs(p.clients) do
        local g = {
            x = area.x,
            y = area.y,
            width = area.width,
            height = area.height
        }
        p.geometries[c] = g
    end
end

--- Maximized layout.
-- @clientlayout awful.layout.suit.max.name
max.name = "max"
function max.arrange(p)
    return fmax(p, false)
end
function max.skip_gap(nclients, t) -- luacheck: no unused args
    return true
end

--- Fullscreen layout.
-- @clientlayout awful.layout.suit.max.fullscreen
max.fullscreen = {}
max.fullscreen.name = "fullscreen"
max.fullscreen.skip_gap = max.skip_gap
function max.fullscreen.arrange(p)
    return fmax(p, true)
end

return max

-- vim: filetype=lua:expandtab:shiftwidth=4:tabstop=8:softtabstop=4:textwidth=80
